### PR TITLE
New bool verification3

### PR DESCRIPTION
**Pull request by josojo**
new isBoolVerification gadget to verify that snark is really operating on bools

### CODE REVIEW

File: pepper/Makefile [hunk 1]
```diff
@@ -43,14 +44,19 @@ link-exo: check-env
...
 e2e-tests: check-env
 	test/e2e/hash_transform.sh
 
-unit-tests: test/apps/hash_transform_test.cpp
-	$(CXX) -DTEST $(TEST_IFLAGS) $< -o $(TEST_BIN)/run_unit_tests $(TEST_LDFLAGS)
+$(FILENAMETESTS)%o: $(FILENAMETESTS)	 
+		$(CXX) $(CXXFLAGS) -DTEST $(TEST_IFLAGS) $(IFLAGS) $< -o $(TEST_BIN)/run_unit_tests $(TEST_LDFLAGS) $(LDFLAGS) 
+		$(TEST_BIN)/run_unit_tests
+
+unit-tests-iBV: test/apps/isBoolVerification_test.cpp
```
Comment (fleupold) [inline pepper/Makefile]: This target is no longer needed?

File: pepper/Makefile [hunk 2]
```diff
@@ -43,14 +44,19 @@ link-exo: check-env
 	ln -sf ~/dex-zksnarks/exo_compute/private_input.py $(PEPPER_BIN)/exo0
 
 # Target to run all tests (e.g. used by travis)
-test: unit-tests gadget-tests e2e-tests
+test: $(FILENAMETESTS)%o gadget-tests e2e-tests
```
Comment (fleupold) [inline pepper/Makefile]: nit: maybe make this a sub-target called unit-tests or app-tests, so it's clear that this is only for those types of files.

File: pepper/test/apps/isBoolVerification_test.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+#include <apps/hash_transform.h>
+
+#include <apps/dex_common.h>
+#include "util.h"
+
+struct TestStructArray { field digest[256]; };
+
+TEST(isBoolVerification, isBoolVerificationWithBools) { 
+    uint32_t length = 2;
+    struct TestStructArray bits[1] = { 0 };
```
Comment (fleupold) [inline pepper/test/apps/isBoolVerification_test.cpp]: Why do you need a struct here? Can't you just do:
```
field  digest[256] = { 0 }
EXPECT_NO_THROW(isBoolVerification(digest, length));
```

File: pepper/test/apps/isBoolVerification_test.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    bits->digest[1] =field("1");
+    bits->digest[2] =field("0");
+    EXPECT_NO_THROW(isBoolVerification(bits->digest, length));
+}
+
+TEST(isBoolVerification, isBoolVerificationWithNonbools) {
+    uint32_t length = 2;
+    struct TestStructArray bits[1]= {0};
```
Comment (fleupold) [inline pepper/test/apps/isBoolVerification_test.cpp]: same here

**Review by fleupold — COMMENTED** [review]
I believe this PR should be based on the one that introduces the new field (rather than master). Otherwise the irrelevant changes show up here as well.

We also need to wait for the pepper PR to go through, right? Otherwise this will not compile on a vanilla pepper build.

File: pepper/test/apps/isBoolVerification_test.cpp (hunk 3 above)
Comment (josojo) [reply pepper/test/apps/isBoolVerification_test.cpp]: yeah... changed it.

**Review by fleupold — APPROVED** [review]

**Comment by josojo** [comment]
New, cleaner PR opened

**Pull request closed without merging**